Build objects

vloginova opened the pull request:
- Rgraph improvements and refactoring
- Build object files for a modular application

vloginova committed:
core: Add (runtime/build)_depends lists

vloginova committed:
core: context: Add to solver rules for module and providers

Each module can provide zero or more modules and must be provided
only once.

vloginova committed:
core: Add Interface(Module)

Interface is a descendant of Module which doesn't provides anything
by default, so that it must me provided by some module.

vloginova committed:
core: Add integer class method to Optype

vloginova committed:
glue: Use 'use' argument only in app tool

Use requires all its members to be built to time the task is run. So
it is not correct to use for all dependents.

vloginova committed:
glue: Add hooks and build_args for diverse inputs

vloginova committed:
glue: Change tools and use full build method

If module doesn't have input sources, standard waf methods cannot
choose necessary features

vloginova committed:
rgraph: Refactor most functions

Refactoring includes workaround with most functions except
Rgraph itself and traverse function.

The only significant change:
	- Function get_error_rgraph() takes is_short argument, True by
	  default. Rgraph and its violation rgraphs are shortened
	  if is_short is true.

vloginova committed:
rgraph: Improve error rgraph traversal

Yields tuples (reason, shift) in the reverse order. which is chosen
in order to make output examination more natural to users.
This method is most useful in case of shortened rgraph.

See traverse_error_graph docs for details

vloginova committed:
rgraph: Refactor rgraph itself

NodeContainer is devided into two classes: Rnode and Container. The
first one contains beacuseof ditionary with container as a key and
reason as a value and dictionry of its containers.
Container has set of its rnode members and therefore dictionary where
rnode is a key and reason is a value.

The rest of refactoring is about workaround and docs.

vloginova committed:
context: Improve provider reasons output

vloginova committed:
glue: mywaf: Create tool and feature for headers generation

vloginova committed:
mywaf: Call ctx.pre_recurse() in an appropriate place

vloginova committed:
glue: Add function for include strings interpolation

vloginova committed:
glue: Add include paths specified by module

vloginova committed:
glue: Change features using in (st/sh)lib taskgens

vloginova committed:
(minor) mywaf: Remove commented code and fix spaces

vloginova committed:
Merge branch 'master' of github.com:vloginova/mybuild into build-objects

review-ninja commented:
<a href="http://app.review.ninja/abusalimov/mybuild/pull/4" target="_blank"><img src="http://app.review.ninja/17731899/pull/4/badge" alt="ReviewNinja"/>

abusalimov commented on glue.py:
```diff
@@ -47,11 +49,19 @@ def options(self, module, ctx):
     def configure(self, module, ctx):
         ctx.load(self.waf_tools)
 
+def interpolate_string(string, env):
```
- [x] This could be implemented through `string.Template(s).substitute(env)`, without messing with regexes.

abusalimov replied:
Plus I would probably move it to `util.misc` package. Or not.

abusalimov commented on glue.py:
```diff
@@ -64,10 +74,27 @@ def define(self, key, val):
+        objects = []
+
+        for fname in module.files:
+            if re.match('.*\.o', fname):
```
- [x] Missing end anchor (otherwise `foo.oww` matches as well). Why not just `fname.endswith('.o')`?

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
         super(CcAppTool, self).build(module, ctx)
-        ctx.program(**self.build_kwargs)
+
+        use = [instance._name for instance in ctx.instance_map \
```
- [x] Avoid line continuation (from [PEP 8](https://www.python.org/dev/peps/pep-0008/#id14)):

> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+        preproc_relative_path = '../../src/{0}/'.format(project_relative_path)
+
+        for fname in module.files:
+            if re.match('.*\.h', fname):
```
- [x] `fname.endswith(...)`

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+            opt_type = 'STRING'
+        elif isinstance(value, bool):
+            opt_type = 'BOOLEAN'
+            value = 1 if value else 0
```
- [x] `value = int(value)`

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+            raise Exception('Option with type {0} is not supported' \
+                            .format(type(value)))
+
+        fmt = 'OPTION_{TYPE}_{MOD}__{NAME} {VALUE}'
```
- [x] Or `return 'OPTION_{opt_type}_{mod_name}__{opt_name} {value}'.format(**locals())`

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+    def get_options(self, module):
+        options = []
+
+        options_dict = module._ModuleBase__optuple.__dict__
```
- [x] Do not ever access private fields!

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+            opt_type = 'NUMBER'
         else:
-            ctx.shlib(**self.build_kwargs)
+            raise Exception('Option with type {0} is not supported' \
```
- [x] Unnecessary line continuation.

abusalimov commented on glue.py:
```diff
@@ -76,38 +103,127 @@ def build(self, module, ctx):
+        options = []
+
+        options_dict = module._ModuleBase__optuple.__dict__
+        for opt_name, opt_val in iteritems(options_dict):
```
- [x] Try this one:

```
for option, value in module._optuple._iterpairs():
    ...
```

abusalimov commented:
I'm going to fix these notes above and merge, although I didn't review rgraph part carefully.

abusalimov committed:
Reimplement interpolate_string through string.Template

abusalimov committed:
glue: Refactor some pieces of code spotted in #4

- Avoid accessing private __optuple field, use _iterpairs() method of
  Optuple instead of getting its __dict__ to iterating over it

- Raise more suitable TypeError on weird option values instead of
  the generic Exception

- Simplify string formatting by using `**locals()` where appropriate,
  namely when there are more than 2-3 format arguments

- Make boolean to number conversion through int(value)
- Remove unused line continuations

abusalimov committed:
rgraph: Make use of Node being iterable yielding its Literals

This also (presumably) avoids creating redundant Rnodes and
overwriting already initialized ones upon Rgraph.__init__.

vloginova commented:
Ok, thank you. You can review rgraph later and I'll happily fix it.

abusalimov committed:
glue: Use str.endswith() to check file extensions instead of re.match
